satellite/satellitedb/usercredits_test.go

```diff
@@ -0,0 +1,178 @@
+// Copyright (C) 2019 Storj Labs, Inc.
+// See LICENSE for copying information.
+
+package satellitedb_test
+
+import (
+	"context"
+	"crypto/rand"
+	"testing"
+	"time"
+
+	"github.com/skyrings/skyring-common/tools/uuid"
+
+	"storj.io/storj/satellite/marketing"
+
+	"github.com/stretchr/testify/require"
+
+	"storj.io/storj/satellite/console"
+
+	"storj.io/storj/internal/testcontext"
+	"storj.io/storj/satellite"
+	"storj.io/storj/satellite/satellitedb/satellitedbtest"
+)
+
+func TestUsercredits(t *testing.T) {
+	t.Parallel()
+
+	satellitedbtest.Run(t, func(t *testing.T, db satellite.DB) {
+		ctx := testcontext.New(t)
+		defer ctx.Cleanup()
+
+		test(ctx, t, db)
+	})
+}
+
+func test(ctx context.Context, t *testing.T, store satellite.DB) {
+	consoleDB := store.Console()
+
+	user, referrer, offer := setupData(ctx, t, store)
+	randomID, err := uuid.New()
+	require.NoError(t, err)
+
+	var inValidUserCredits = []console.UserCredit{
+		{
+			UserID:               *randomID,
+			OfferID:              offer.ID,
+			ReferredBy:           referrer.ID,
+			CreditsEarnedInCents: 100,
+			ExpiresAt:            time.Now().UTC().AddDate(0, 1, 0),
+		},
+		{
+			UserID:               user.ID,
+			OfferID:              10,
+			ReferredBy:           referrer.ID,
+			CreditsEarnedInCents: 100,
+			ExpiresAt:            time.Now().UTC().AddDate(0, 1, 0),
+		},
+		{
+			UserID:               user.ID,
+			OfferID:              offer.ID,
+			ReferredBy:           *randomID,
+			CreditsEarnedInCents: 100,
+			ExpiresAt:            time.Now().UTC().AddDate(0, 1, 0),
+		},
+	}
+
+	for _, ivc := range inValidUserCredits {
+		_, err := consoleDB.UserCredits().Create(ctx, ivc)
+		require.Error(t, err)
+	}
+
+	var validUserCredits = []struct {
+		userCredit     console.UserCredit
+		chargedCredits int
+		expected       int
+	}{
+		{
+			userCredit: console.UserCredit{
+				UserID:               user.ID,
+				OfferID:              offer.ID,
+				ReferredBy:           referrer.ID,
+				CreditsEarnedInCents: 100,
+				ExpiresAt:            time.Now().UTC().AddDate(0, 1, 0),
+			},
+			chargedCredits: 120,
+			expected:       0,
+		},
+		{
+			userCredit: console.UserCredit{
+				UserID:               user.ID,
+				OfferID:              offer.ID,
+				ReferredBy:           referrer.ID,
+				CreditsEarnedInCents: 100,
+				ExpiresAt:            time.Now().UTC().AddDate(0, 0, -5),
+			},
+			chargedCredits: 60,
+			expected:       0,
```

Comment:
similarly, I think expected should be nonzero here as well. Seems like it should be 40 (if I understand how this works, which I might not)

Reply:
I added a comment for this entry to highlight that it should not be applied since the `expires_at` is set to be before `Now()`